lib/open_food_network/sales_tax_report.rb

```diff
@@ -24,19 +42,38 @@ def orders
     end
 
     def table
-      orders.map do |order|
-        totals = totals_of order.line_items
-        shipping_cost = shipping_cost_for order
-
-        [order.number, order.created_at, totals[:items], totals[:items_total],
-         totals[:taxable_total], totals[:sales_tax], shipping_cost, order.shipping_tax, order.enterprise_fee_tax, order.total_tax,
-         order.bill_address.full_name, order.distributor.andand.name]
+      case params[:report_type]
+      when "tax_types"
+        orders.map do |order|
+          totals = totals_of order.line_items
+          shipping_cost = shipping_cost_for order
+
+          [order.number, order.created_at, totals[:items], totals[:items_total],
+           totals[:taxable_total], totals[:sales_tax], shipping_cost, order.shipping_tax, order.enterprise_fee_tax, order.total_tax,
+           order.bill_address.full_name, order.distributor.andand.name]
+        end
+      else
+        orders.map do |order|
+          [order.number, order.total - order.total_tax] +
+            relevant_rates.map { |rate| order.tax_adjustment_totals.fetch(rate, 0) } +
+            [order.total_tax, order.display_total]
+        end
       end
     end
 
 
     private
 
+    def relevant_rates
+      queries = [ search.result.joins(:line_items => {:adjustments => :tax_rate}).select('spree_tax_rates.*').uniq,
+                  search.result.joins(:adjustments => :tax_rate).select('spree_tax_rates.*').uniq ]
+      queries.map do |query|
+        ActiveRecord::Base.connection.select_all(query)
+      end.sum.map do |tax_rate|
+        Spree::TaxRate.new(tax_rate, without_protection: true)
+      end
+    end
+
```

Comment:
Ok, so I think I now understand this a bit better. Is there any particular reason that you need new instances of Spree::TaxRate instead of just using the ones from the db? Could we could re-write this function? How about:

```ruby
def relevant_rates
  return @relevant_rates unless @relevant_rates.nil?
  item_rate_ids = search.result.joins(:line_items => {:adjustments => :tax_rate}).select('spree_tax_rates.id').uniq
  order_rate_ids = search.result.joins(:adjustments => :tax_rate).select('spree_tax_rates.id').uniq
  @relevant_rates = Spree::TaxRate.where(id: item_rate_ids | order_rate_ids)
end
```

Sure, it is three queries rather than two, but I think it is more readable and we could make it one query by using subqueries if you were particularly worried about the number of queries. I think the important thing is that we are not re-running all three queries for every order that we pull up, it just runs once and stores the result for later use.

Perhaps you are right - maybe we should just display all available rates?